File: hard/41_first_missing_positive.py
from typing import List

class Solution:
    def firstMissingPositive(self, nums: List[int]) -> int:
        # missing positive integer of a array of size N can range from 1 to N+1
        N = len(nums)
        # remove the values which dont consititute in the final answer
        # that 0 and values > N+1
        # we do that by putting max value N+1 in place of those values
        for i in range(N):
            if nums[i] <= 0 or nums[i] > N:
                nums[i] = N+1

        # mark the index
        for i in range(N):
            num = abs(nums[i])
            # mark that this number is already seen so it cant be the answer
            if num > N:
                continue
            num -= 1
            if nums[num] > 0:
                nums[num] = -1*nums[num]

        for i in range(N):
            if nums[i] >= 0:
                return i+1

        return N+1


if __name__ == "__main__":
    solution = Solution()
    assert solution.firstMissingPositive([3,4,-1,1]) == 2

File: hard/41_first_missing_positive.rs
impl Solution {
    pub fn first_missing_positive(nums_original: Vec<i32>) -> i32 {
        // take a copy
        let mut nums = nums_original.clone();
        let N = nums.len() as i32;
        for (index, num) in nums_original.iter().enumerate() {
            if num < &1 || num > &N {
                nums[index] = N+1;
            }
        }
        // mark the index as visited by making the value negative
        for index in 0..nums.len() {
            let mut value = nums[index].abs() as usize;
            if value > N as usize {
                continue;
            }
            value = value - 1;
            if nums[value] > 0 {
                nums[value] = -1 * nums[value] as i32;
            }
        }
        println!("{:?}", nums);
        // get the first index, wich is having positive value
        for index in 0..nums.len() {
            if nums[index] > 0 {
                return index as i32 + 1;
            }
        }
        N+1
    }
}
